addons/enemybar/actionTracking.lua

```diff
@@ -0,0 +1,299 @@
+res = require('resources')
+vector = require('vectors')
+packets = require('packets')
+
+tracked_actions = {}
+tracked_enmity = {}
+tracked_debuff = {}
+framerate = 75
+clean_actions_delay = framerate
+clean_actions_tick = clean_actions_delay
+
+function looking_at(a, b)
+	local h = a.facing % math.pi
+	local h2 = (math.atan2(a.x-b.x,a.y-b.y) + math.pi/2) % math.pi
+	return math.abs(h-h2) < 0.15
+end
+
+function is_party_member_or_pet(mob)
+	if not mob then return false end
+	if mob.id == windower.ffxi.get_player().id then return true end
+	local is_pet = mob.is_npc and mob.charmed
+
+	local party = windower.ffxi.get_party()
+	for i=0, party.party1_count-1 do
+		local p = party['p'..i]
+		if p and p.mob then
+			if is_pet then
+				if p.mob.pet_index == mob.index then return true end
+			else
+				if p.mob.id == mob.id then return true end
+			end
+		end
+	end
+	for i=0, party.party2_count-1 do
+		local p = party['a1'..i]
+		if p and p.mob then
+			if is_pet then
+				if p.mob.pet_index == mob.index then return true end
+			else
+				if p.mob.id == mob.id then return true end
+			end
+		end
+	end
+	for i=0, party.party3_count-1 do
+		local p = party['a2'..i]
+		if p and p.mob then
+			if is_pet then
+				if p.mob.pet_index == mob.index then return true end
+			else
+				if p.mob.id == id then return true end
+			end
+		end
+	end
+
+    return false
+end
+
+function handle_action_packet(id, data)
+	if 0x028 == id then
+		local ai = windower.packets.parse_action(data)
+
+		track_enmity(ai)
+		track_actions(ai)
+		track_debuffs(ai)
+	elseif 0x029 then
+    	local message_id = data:unpack('H',0x19)
+    	if not message_id then return end
```

Comment:
This check can be removed, it never fails (you have a habit of overchecking things throughout the addon, such as checking for `p` above, and a few more times below, I will try to mention it if I think of it).

Reply:
Actually, I added this in testing. It was actually failing sometimes without that check. The packet library may have been returning a nil packet because it didn't read properly.